### ISSUE TITLE
egrep: warning: egrep is obsolescent; using grep -E

### ISSUE THREAD

**Issue by Obema12** [CISOfy/lynis#1338]
The execution of `lynis audit system` is accompanied by a whole bunch of messages related to grep
```
egrep: warning: egrep is obsolescent; using grep -E
  - Total without nodev:8 noexec:13 nosuid:6 ro or noexec (W^X): 13 of total 27
  - Disable kernel support of some filesystems
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
```

Fix it at last.

**Comment by mboelen** [comment]
On what operating system and version are you running Lynis?

**Comment by Obema12** [comment]
Archlinux.

**Comment by konstruktoid** [comment]
man grep(1) on Ubuntu 20.04.5:
"the variant programs egrep, fgrep and rgrep are the same as grep -E, grep -F, and grep -r, respectively.  These variants are deprecated, but are provided for backward compatibility."

**Comment by mboelen** [comment]
Looks like we have some conversion to do. The reason to use `egrep` was that `grep` on Solaris did not support the `-E` flag. A shame that the command is spewing messages now and therefore breaking that backward compatibility.

@Obema12 Thanks for your feedback. Next time when you report an issue, please leave off the "Fix it at last." at the end of your issue. It sounds unfriendly, while it may or may not be intended.

**Comment by konstruktoid** [comment]
Just to clarify; `egrep`, `fgrep` or `rgrep` don't spew any messages regarding deprecation on Ubuntu.

**Comment by mboelen** [comment]
Yup. Looks like someone decided to turn on this message in a recent version of the package that contains the related utilities. Similar messages show up, most of them are from this month. With Arch Linux using often the newest packages it is no surprise it pops up there first. So it definitely makes sense to convert now that the compatibility starts to break.

**Comment by Obema12** [comment]
@mboelen tried to be polite. But dozens of problems found and sent show that politeness does not achieve anything.

I sent rougher questions and it got a much faster developer response. It got to the point that rudeness, published immediately after politeness, was decided immediately! Whereas the polite message remained open for months.

I don't know how it works.

**Comment by mboelen** [comment]
So it was intended. In that case, let me be very clear: we have a [Code of Conduct](https://github.com/CISOfy/lynis/blob/master/CODE_OF_CONDUCT.md) (listed here on GitHub, including under the comment form "Remember, contributions to this repository...").

The CoC states that we expect people to be welcoming and friendly within this community, and also my duties as a project owner. It would be nice to treat people how you want to be treated yourself.

If you can't (or won't like to) accept these rules within the CoC, please don't be part of our community.

**Comment by fw-ops-automation-imprivata** [comment]
There are countless executeables that behave differently according to how they are named. This whole deprecated thing is developers who can't find something useful to do. 30+ years of scripts with egrep and fgrep. If Linux folks keep this stuff up it will eventually be as nasty as Wyndoze. 
If It Ain't Broke, Don't "Fix" it. Pure pedantry.

**Comment by fw-ops-automation-imprivata** [comment]
I guess I can set up a shell alias. :|

**Comment by webcaptcha** [comment]
@fw-ops-automation-imprivata were you success in writing that alias? I am looking for workaround too.

**Comment by OckhamOdyssey** [comment]
@webcaptcha I tried right now. I add the alias in the bashrc, zshrc, root bashrc and root zshrc. Nothing worked :/

**Comment by mlude** [comment]
the warning message was introduced in gnu grep 3.8

**Comment by jokeyrhyme** [comment]
> The reason to use egrep was that grep on Solaris did not support the -E flag.

Could we do a one-time test at the very beginning that checks whether `grep -E` works, and if it does not then fallback to `egrep` ?

**Comment by konstruktoid** [comment]
agreed @jokeyrhyme and then replace `${EGREPBINARY} with a working command, so it doesn't need such a big rewrite.

**Comment by wileyhy** [comment]
This bug is still an issue.

** # `lynis --version`
egrep: warning: egrep is obsolescent; using grep -E
egrep: warning: egrep is obsolescent; using grep -E
3.0.8

**Comment by wileyhy** [comment]
Note: 

** # `diff -u /usr/share/lynis/include/functions_orig /usr/share/lynis/include/functions`

> --- /usr/share/lynis/include/functions_orig	2022-05-16 17:00:00.000000000 -0700
> +++ /usr/share/lynis/include/functions	2023-06-22 09:01:05.372876172 -0700
> @@ -670,6 +670,19 @@
>  
>  
>      ################################################################################
> \+    \# Name        : egrep (alias)
> \+    \# Description : eliminate the errors while still being usable for older systems
> \+    \#    
> \+    \# Parameters  : $1 = text
> \+    \# Returns     : <nothing>
> \+    \################################################################################
> \+
> \+    if egrep 2>&1 | grep -q 'egrep: warning: egrep is obsolescent; using grep -E'; then
> \+      alias egrep='grep -E'
> \+    fi
> \+
> \+
> \+    \################################################################################
>      \# Name        : Equals()
>      \# Description : Compare two strings after special characters were stripped
>      \#
>

**Comment by wileyhy** [comment]
`egrep` isn't POSIX any more, as it happens.

> The Open Group Base Specifications Issue 7, 2018 edition
> IEEE Std 1003.1-2017
> ...
> C.4.2  Utilities Removed in the Previous Version
> The following utilities were removed in the previous version of this standard:
> ...egrep fgrep..
> 

[https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_xcu_chap04.html
](url)

**Comment by xnoguer** [comment]
@mboelen This issue can probably be closed now after these commits:
https://github.com/CISOfy/lynis/commit/e816df538fee086927bb12d97eec16532baeb4c0
https://github.com/CISOfy/lynis/commit/1dad487318f36398b3873c07ac157aa7e3d75351

**Comment by xusism** [comment]
> [@mboelen](https://github.com/mboelen) This issue can probably be closed now after these commits: [e816df5](https://github.com/CISOfy/lynis/commit/e816df538fee086927bb12d97eec16532baeb4c0) [1dad487](https://github.com/CISOfy/lynis/commit/1dad487318f36398b3873c07ac157aa7e3d75351)

Hi, I'm still getting this issue in OpenSUSE (version 3.1.4)

**Comment by aztrix** [comment]
Hi Folks

I'm also getting this error on openSUSE Tumbleweed-Slowroll  x86_64/Lynis 3.1.4

Question, if I leave it running will it finish eventually or is it in an endless loop?

Please let me know if you require additional diagnostics?

Regards
Aztrix

**Comment by xusism** [comment]
> I'm also getting this error on openSUSE Tumbleweed-Slowroll x86_64/Lynis 3.1.4

Okay, it seems we should use the Git method instead of the zypper package, which uses Lynis 3.1.5 and works perfectly: https://github.com/CISOfy/lynis?tab=readme-ov-file#git

**Comment by mboelen** [comment]
This issue should be resolved, so will close this issue. A new packaged release will be made available soon.
